docs/weave/design/04-queries.md

```diff
@@ -0,0 +1,96 @@
+
+```go
+type ResponseQuery struct {
+    Code uint32
+    Log    string
+    Info   string
+    Index  int64
+    Key    []byte
+    Value  []byte
+    Proof  []byte
+    Height int64
+}
+```
+
+That's a lot of fields... let's skip through them. `Code` is set to non-zero only when there is an error in processing the query. `Log` and `Info` are human readable strings for debugging or extra info. `Index` *may* be the location of this key in the merkle tree, but that is not well defined.
+
+Now to the important ones. `Height`, as above, the the block height of the tree we queried and is always set, even if the query had 0 to request "most recent". `Key` is the key of the merkle tree we got, `Value` is the value stored at that key (may be empty if there is no value), and `Proof` is a merkle proof in an undefined format.
+
+## Weave Request Types
+
+As we see above, the request format doesn't actually define what possible types are for either `Path` or `Data` and leaves it up to the application. This is good for a generic query interface, but to allow better code reuse between weave extensions, as well as ease of development of weave clients, we define a standard here for all weave modules.
+
+### Constructing Paths
+
+Paths includes the resource we want to get:
+
+- Raw Key: `/`
+- Bucket: `/[bucket]`
+- Index: `/[bucket]/[index]`
+
+By default, we expect `Data` to include a raw key to match in that context. However, we can also append a modifier to change that behavior:
+
+- `?prefix` =\> `Data` is a raw prefix (query returns N results, all items that start with this prefix)
+
+### Examples
+
+[namecoin.NewWalletBucket](https://github.com/iov-one/weave/blob/master/x/namecoin/wallet.go#L107-L113) adds and a `name` field to the account, along with a secondary index. It is [registered under /wallets](https://github.com/iov-one/weave/blob/master/x/namecoin/handler.go#L52-L57) in the QueryHandler.
+
+Path: `/`, Data: `0123456789` (hex):  
+db.Get(`0123456789`)
+
+Path: `/wallets`, Data: `00CAFE00` (hex):  
+namecoin.NewWalletBucket().Get(`00CAFE00`)
+
+Path: `/wallets/name`, Data: "John" (raw):  
+namecoin.NewWalletBucket().Index("name").Get("John")
+
+Path: `/?prefix`, Data: `0123456789` (hex):  
+db.Iterator(`0123456789`, `012345678A`)
+
+### Weave Response Types
+
+Some queries return single responses, others multiple. Rather that some complex switch statement in either the client or the application, the simplest approach is to learn from other databases, and always return a `ResultSet`. A higher-level client wrapper can provide nicer interfaces, but this provides a consistent format they can build on.
```

Comment:
"Rather that some complex switch statement in either the client or the application, the simplest approach is to learn from other databases, and always return a"

change to 

"Rather than build a complex switch statement in either the client or the application, the simplest approach is to learn from other databases, and always return a"